arangod/Cluster/RebootTracker.h

```diff
@@ -23,119 +23,83 @@
 class Slice;
-}  // namespace velocypack
 
+}  // namespace velocypack
 namespace cluster {
 
 // Note:
 // Instances of this class must be destructed during shutdown before the
 // scheduler is destroyed.
 class RebootTracker {
  public:
-  using Callback = std::function<void(void)>;
   using SchedulerPointer = SupervisedScheduler*;
-
-  class PeerState {
-   public:
-    PeerState(ServerID serverId, RebootId rebootId)
-        : _serverId{std::move(serverId)}, _rebootId{rebootId} {}
-
-    ServerID const& serverId() const noexcept { return _serverId; }
-    RebootId rebootId() const noexcept { return _rebootId; }
-
-    void toVelocyPack(velocypack::Builder& builder) const;
-    static PeerState fromVelocyPack(velocypack::Slice);
-
-   private:
-    ServerID _serverId;
-    RebootId _rebootId;
+  using Callback = fu2::unique_function<void()>;
+  struct DescriptedCallback {
+    Callback callback;
+    std::string description;
   };
+  using State = containers::FlatHashMap<ServerID, RebootId>;
 
   explicit RebootTracker(SchedulerPointer scheduler);
 
-  CallbackGuard callMeOnChange(PeerState const& peerState, Callback callback,
-                               std::string callbackDescription);
+  CallbackGuard callMeOnChange(std::string_view serverId, RebootId rebootId,
+                               Callback callback, std::string_view description);
 
-  void updateServerState(
-      containers::FlatHashMap<ServerID, RebootId> const& state);
+  void updateServerState(State state);
 
  private:
   using CallbackId = uint64_t;
+  using CallbackIds = containers::FlatHashMap<CallbackId, DescriptedCallback>;
+  using RebootIds = std::map<RebootId, CallbackIds>;
+  using Callbacks = containers::FlatHashMap<ServerID, RebootIds>;
+  using CallbacksIt = Callbacks::iterator;
 
-  struct DescriptedCallback {
-    Callback callback;
-    std::string description;
-  };
-
-  CallbackId getNextCallbackId() noexcept;
-
-  void unregisterCallback(PeerState const&, CallbackId);
-
-  // bool notifyChange(PeerState const& peerState);
-
-  void scheduleAllCallbacksFor(ServerID const& serverId);
-  void scheduleCallbacksFor(ServerID const& serverId, RebootId rebootId);
+  void unregisterCallback(std::string_view serverId, RebootId rebootId,
+                          CallbackId callbackId) noexcept;
 
-  static Callback createSchedulerCallback(
-      std::vector<
-          std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-          callbacks);
+  void queueCallback(DescriptedCallback&& callback) noexcept;
+  void queueCallbacks(CallbacksIt it) noexcept;
+  void queueCallbacks(CallbacksIt it, RebootId to);
 
-  void queueCallbacks(
-      std::vector<
-          std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-          callbacks);
-  void queueCallback(DescriptedCallback callback);
+  mutable std::mutex _mutex;
 
- private:
-  mutable Mutex _mutex;
+  CallbackId _callbackId;
 
-  CallbackId _nextCallbackId{1};
```

Comment:
I think we have 2 points:
1. call ctor in cpp or header
I don't agree with you (we have compiler for check this), but It's not important for me, so I can revert it
2. I wrote `_callbackId`, `{0}`, `operator++()` instead of `_nextCallbackId`, `{1}`, `operator++(int)`
I made it because I thought it a little simpler, but I also can revert it, if you want

Reply:
1. ✓

2. Yes, the current code is slightly simpler. The reasoning for the previous implementation was a different one, though: Namely that `_nextCallbackId` immediately conveys its behavior, without looking at any code: That it's the next ID of something, and that it's not yet used itself. While `_callbackId` conveys nothing but its type. Could be the ID of some special callback. Could also be the next ID. Or something else. And as I didn't have an equally expressive name for the current variant, I chose to go with one additional LOC in favor of having a better name.

Reply:
2. https://github.com/arangodb/arangodb/pull/16375/commits/3d7839382e569cea9bc4be5bd65c5faf322994b0
Done?